src/c++/perf_analyzer/client_backend/triton_c_api/shared_memory_manager.cc

```diff
@@ -0,0 +1,184 @@
+    return Error(
+        std::string("shared memory region '" + name + "' already in manager"));
+  }
+
+  shared_memory_map_.insert(std::make_pair(
+      name, std::unique_ptr<MemoryInfo>(new MemoryInfo(
+                name, 0 /* offset */, byte_size, ptr, TRITONSERVER_MEMORY_CPU,
+                0 /* device id */))));
+
+  return Error::Success;
+}
+
+Error
+SharedMemoryManager::GetMemoryInfo(
+    const std::string& name, size_t offset, void** shm_mapped_addr,
+    TRITONSERVER_MemoryType* memory_type, int64_t* device_id)
+{
+  // protect shared_memory_map_ from concurrent access
+  std::lock_guard<std::mutex> lock(mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it == shared_memory_map_.end()) {
+    return Error(
+        std::string("Unable to find shared memory region: '" + name + "'"));
+  }
+  if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+    *shm_mapped_addr =
+        (void*)((uint8_t*)it->second->mapped_addr_ + it->second->offset_ + offset);
+  } else {
+    *shm_mapped_addr = (void*)((uint8_t*)it->second->mapped_addr_ + offset);
+  }
+
+  *memory_type = it->second->kind_;
+  *device_id = it->second->device_id_;
+
+  return Error::Success;
+}
+
+
+Error
+SharedMemoryManager::Unregister(
+    const std::string& name, TRITONSERVER_MemoryType memory_type)
+{
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(mu_);
+
+  return UnregisterHelper(name, memory_type);
+}
+
+Error
+SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+  std::string error_message = "Failed to unregister the following ";
+  std::vector<std::string> unregister_fails;
+  if (memory_type == TRITONSERVER_MEMORY_CPU) {
+    // Serialize all operations that write/read current shared memory regions
+    error_message += "system shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
+         it != shared_memory_map_.cend(); it = next_it) {
+      ++next_it;
+      if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+        Error err = UnregisterHelper(it->first, memory_type);
+        if (!err.IsOk()) {
+          unregister_fails.push_back(it->first);
+        }
+      }
+    }
+  } else if (memory_type == TRITONSERVER_MEMORY_GPU) {
+    error_message += "cuda shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
+         it != shared_memory_map_.cend(); it = next_it) {
+      ++next_it;
+      if (it->second->kind_ == TRITONSERVER_MEMORY_GPU) {
+        Error err = UnregisterHelper(it->first, memory_type);
+        if (!err.IsOk()) {
+          unregister_fails.push_back(it->first);
+        }
+      }
+    }
+  }
+
+  if (!unregister_fails.empty()) {
+    for (auto unreg_fail : unregister_fails) {
+      error_message += unreg_fail + " ,";
+    }
+    std::cerr << error_message << std::endl;
```

Comment:
Why print and return the error object with the same message? Wouldn't the responsibility of informing the user of the error message be whatever higher up function receives this error object?